src/librustc/infer/const_variable.rs

```diff
@@ -0,0 +1,271 @@
+use crate::mir::interpret::ConstValue;
+use syntax::symbol::InternedString;
+use syntax_pos::Span;
+use crate::ty::{self, InferConst};
+
+use std::cmp;
+use std::marker::PhantomData;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_data_structures::snapshot_vec as sv;
+use rustc_data_structures::unify as ut;
+
+pub struct ConstVariableTable<'tcx> {
+    values: sv::SnapshotVec<Delegate<'tcx>>,
+
+    relations: ut::UnificationTable<ut::InPlace<ty::ConstVid<'tcx>>>,
+}
+
+/// Reasons to create a const inference variable
+#[derive(Copy, Clone, Debug)]
+pub enum ConstVariableOrigin {
+    MiscVariable(Span),
+    ConstInference(Span),
+    ConstParameterDefinition(Span, InternedString),
+    SubstitutionPlaceholder(Span),
+}
+
+pub type ConstVariableMap<'tcx> = FxHashMap<ty::ConstVid<'tcx>, ConstVariableOrigin>;
+
+struct ConstVariableData {
+    origin: ConstVariableOrigin,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum ConstVariableValue<'tcx> {
+    Known { value: &'tcx ty::LazyConst<'tcx> },
+    Unknown { universe: ty::UniverseIndex },
```

Comment:
Is this enough? Don't we also need the type?

Reply:
Well, I think the type is carried in the `ty::Const` wrapper, for better or worse:

```rust
    #[inline]
    pub fn mk_const_var(self, v: ConstVid<'tcx>, ty: Ty<'tcx>) -> &'tcx Const<'tcx> {
        self.mk_const(ty::Const {
            val: ConstValue::Infer(InferConst::Var(v)),
            ty,
        })
    }
```

Reply:
It's not carried around in `Unknown`, which I think is the part @eddyb was referring to. (There's some other discussion about this elsewhere, but there's a `FIXME` in the code about this now.)